subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,531 @@
+
+	if (sst->year < 1582 || sst->year > 9999) {
+		LOG_WRN("Invalid SST year");
+		return false;
+	}
+	if (sst->month < 1 || sst->month > 12) {
+		LOG_WRN("Invalid SST month");
+		return false;
+	}
+	if (leap_year_check(sst->year)) {
+		if (sst->day < 1 || sst->day > day_of_month_leap[sst->month]) {
+			LOG_WRN("Invalid SST day");
+			return false;
+		}
+	} else {
+		if (sst->day < 1 || sst->day > day_of_month[sst->month]) {
+			LOG_WRN("Invalid SST day");
+			return false;
+		}
+	}
+	if (sst->hour > 23) {
+		LOG_WRN("Invalid SST hour");
+		return false;
+	}
+	if (sst->minute > 59) {
+		LOG_WRN("Invalid SST minute");
+		return false;
+	}
+	if (sst->second > 59) {
+		LOG_WRN("Invalid SST second");
+		return false;
+	}
+	if (!(sst->timezone == -128 || (sst->timezone >= -48 && sst->timezone <= 56))) {
+		LOG_WRN("Invalid SST timezone");
+		return false;
+	}
+	if (!(sst->dst_offset == 0 || sst->dst_offset == 2 || sst->dst_offset == 4
+		|| sst->dst_offset == 8 || sst->dst_offset == 255)) {
+		LOG_WRN("Invalid SST DST offset");
+		return false;
+	}
+	return true;
+}
+
+static int cal_sst_offset(struct net_buf_simple *sst_buffer)
+{
+	struct cgms_session_start_time sst;
+	uint32_t time_offset;
+	uint32_t day_to_change;
+	uint32_t second_to_change;
+	uint32_t residual_day;
+	uint32_t residual_second;
+	int day_of_year[13] = {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365};
+	int day_of_leap_year[13] = {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366};
+
+	if (!sst_buffer || sst_buffer->len < CGMS_SST_LENGTH) {
+		return -EINVAL;
+	}
+
+	sst.year = net_buf_simple_pull_le16(sst_buffer);
+	sst.month = net_buf_simple_pull_u8(sst_buffer);
+	sst.day = net_buf_simple_pull_u8(sst_buffer);
+	sst.hour = net_buf_simple_pull_u8(sst_buffer);
+	sst.minute = net_buf_simple_pull_u8(sst_buffer);
+	sst.second = net_buf_simple_pull_u8(sst_buffer);
+	sst.timezone = net_buf_simple_pull_u8(sst_buffer);
+	sst.dst_offset = net_buf_simple_pull_u8(sst_buffer);
+
+	if (!valid_datetime_check(&sst)) {
+		LOG_WRN("Invalid SST");
+		return -EINVAL;
+	}
+
+	/* Get elapsed time in second */
+	time_offset = (k_uptime_get_32() - cgms_inst.local_start_time) / MSEC_PER_SEC;
+	day_to_change = time_offset / SEC_PER_DAY;
+	second_to_change = time_offset % SEC_PER_DAY;
+	residual_second = sst.hour * 3600 + sst.minute * 60 + sst.second;
+	if (residual_second < second_to_change) {
+		residual_second += SEC_PER_DAY;
+		day_to_change++;
+	}
+	/* hh:mm:ss */
+	sst.hour = (residual_second - second_to_change) / 3600;
+	sst.minute = (residual_second - second_to_change - sst.hour * 3600) / 60;
+	sst.second = residual_second - second_to_change - sst.hour * 3600 - sst.minute * 60;
+
+	/* yyyy:mm:dd */
+	residual_day = leap_year_check(sst.year) ?
+		day_of_leap_year[sst.month - 1] : day_of_year[sst.month - 1];
+	residual_day += sst.day;
+
+	while ((residual_day - day_to_change) <= 0) {
+		residual_day += leap_year_check(sst.year - 1) ? 366 : 365;
+		sst.year--;
```

Comment:
Considering the range of `k_uptime_get_32()`, it seems to me that we are not able to decrease the `sst.year`.

Reply:
The maximum value of time offset would be 2^32 miliseconds = 49.7 days. I think it is large enough in practice. It is still possible to reduce the year. For example, if the session lasts 12 hours already, and the user enters a time of 2022-01-01 10:00:00, then the CGMS needs to shift the year and gives a session start time of 2021-12-31 22:00:00.

Reply:
You're right, thanks for the clarification.